src/parser/tokenizer.ts

```diff
@@ -191,6 +192,24 @@ export class Tokenizer {
     throw this.mkError(`raw ${this.snapshot(this.rawBeginAt)} not closed`, begin)
   }
 
+  readLiquidTagTokens (options: NormalizedFullOptions = defaultOptions): LiquidTagToken[] {
+    const tokens: LiquidTagToken[] = []
+    while (this.p < this.N) {
+      const token = this.readLiquidTagToken(options)
+      if (token.name) tokens.push(token)
```

Comment:
Here, `if (token.name) tokens.push(token)` is handling the case of an empty line mid `liquid` tag. As implemented, `token` is always a `LiquidTagToken`, with a `kind` of `TokenKind.Tag`, only sometimes and empty one.

An alternative solution I considered was to check for whitespace only lines in `readLiquidTagToken`. While this did avoid unnecessarily creating an "empty" token, it would involve slicing the input twice for non-empty tokens (once in `readLiquidTagToken` and once in the token constructor) and potentially returning something other than a token from `readLiquidTagToken`. Making line 99 `if (token !== undefined) tokens.push(token)` instead.

```typescript
readLiquidTagToken (options: NormalizedFullOptions): LiquidTagToken | undefined {
  const { file, input } = this
  const begin = this.p
  let end = this.N
  if (this.readToDelimiter('\n') !== -1) end = this.p
  if (!/\S/.test(input.slice(begin, end))) return undefined
  return new LiquidTagToken(input, begin, end, options, file)
}
```

Checking for an empty token name seemed less bad to me.

Reply:
Sounds reasonable!